diff --git a/apps/system/js/sim_lock_manager.js b/apps/system/js/sim_lock_manager.js
index 1c30748..1f8bd70 100644
--- a/apps/system/js/sim_lock_manager.js
+++ b/apps/system/js/sim_lock_manager.js
@@ -162,6 +162,20 @@
       // https://bugzilla.mozilla.org/show_bug.cgi?id=SIMPIN-Dialog
     },
 
+    isBothSlotsLocked: function sl_isBothSlotsLocked(){   
+      if(!SIMSlotManager.isMultiSIM() ||   
+          SIMSlotManager.hasOnlyOneSIMCardDetected()){
+        return false;
+      }
+
+      var simSlots = SIMSlotManager.getSlots();
+      var isBothLocked = true;
+      for (var i = 0; i < simSlots.length; i++) {
+        isBothLocked = isBothLocked && simSlots[i].isLocked(); 
+      };
+      return isBothLocked;  
+    },
+
     showIfLocked: function sl_showIfLocked(currentSlotIndex, skipped) {
       if (!SIMSlotManager.ready) {
         this.warn('SIMSlot not ready yet.');
@@ -219,8 +233,7 @@
         }
 
         // Always showing the first slot first.
-        if (!this._alreadyShown &&
-          !SIMSlotManager.hasOnlyOneSIMCardDetected() && index > 0) {
+        if (!this._alreadyShown && this.isBothSlotsLocked() && index > 0) {
           return false;
         }
 
diff --git a/apps/system/test/unit/sim_lock_manager_test.js b/apps/system/test/unit/sim_lock_manager_test.js
index ff61486..4295c21 100644
--- a/apps/system/test/unit/sim_lock_manager_test.js
+++ b/apps/system/test/unit/sim_lock_manager_test.js
@@ -257,6 +257,34 @@ suite('SimLockManager', function() {
           assert.equal(subject.simLockSystemDialog.show.callCount, 1);
       });
     });
+
+    suite('only second slot is locked', function() {
+      function removeSIMLockForFirstSlot() {
+        MockSIMSlotManager.mInstances[0].simCard.cardState = 'ready';
+      }
+
+      function restoreSIMLockForFirstSlot() {
+        MockSIMSlotManager.mInstances[0].simCard.cardState = 'pinRequired';
+      }
+
+      setup(function() {
+        MockSIMSlotManager.ready = true;
+        addSimSlot();
+        removeSIMLockForFirstSlot();
+        subject._alreadyShown = false;
+      });
+
+      teardown(function() {
+        removeSimSlot();
+        restoreSIMLockForFirstSlot();
+      });
+
+      test('should show the second SIM if the first is not locked',
+        function() {
+          subject.showIfLocked();
+          assert.equal(subject.simLockSystemDialog.show.callCount, 1);
+      });
+    });
   });
 
   suite('lockscreen request to unlock', function() {
