// chai/lib-cov/error.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['error.js']) {
  _$jscoverage['error.js'] = [];
  _$jscoverage['error.js'][7] = 0;
  _$jscoverage['error.js'][13] = 0;
  _$jscoverage['error.js'][14] = 0;
  _$jscoverage['error.js'][15] = 0;
  _$jscoverage['error.js'][16] = 0;
  _$jscoverage['error.js'][17] = 0;
  _$jscoverage['error.js'][18] = 0;
  _$jscoverage['error.js'][19] = 0;
  _$jscoverage['error.js'][21] = 0;
  _$jscoverage['error.js'][22] = 0;
  _$jscoverage['error.js'][26] = 0;
  _$jscoverage['error.js'][28] = 0;
  _$jscoverage['error.js'][29] = 0;
}
_$jscoverage['error.js'][7]++;
module.exports = AssertionError;
_$jscoverage['error.js'][13]++;
function AssertionError(options) {
  _$jscoverage['error.js'][14]++;
  options = options || {};
  _$jscoverage['error.js'][15]++;
  this.name = "AssertionError";
  _$jscoverage['error.js'][16]++;
  this.message = options.message;
  _$jscoverage['error.js'][17]++;
  this.actual = options.actual;
  _$jscoverage['error.js'][18]++;
  this.expected = options.expected;
  _$jscoverage['error.js'][19]++;
  this.operator = options.operator;
  _$jscoverage['error.js'][21]++;
  if (Error.captureStackTrace) {
    _$jscoverage['error.js'][22]++;
    Error.captureStackTrace(this, options.stackStartFunction);
  }
}
_$jscoverage['error.js'][26]++;
AssertionError.prototype.__proto__ = Error.prototype;
_$jscoverage['error.js'][28]++;
AssertionError.prototype.toString = (function () {
  _$jscoverage['error.js'][29]++;
  return this.message;
});
_$jscoverage['error.js'].source = ["/*!"," * chai"," * Copyright(c) 2011 Jake Luer &lt;jake@alogicalparadox.com&gt;"," * MIT Licensed"," */","","module.exports = AssertionError;","","/*!"," * Inspired by node.js assert module"," * https://github.com/joyent/node/blob/f8c335d0caf47f16d31413f89aa28eda3878e3aa/lib/assert.js"," */","function AssertionError (options) {","  options = options || {};","  this.name = 'AssertionError';","  this.message = options.message;","  this.actual = options.actual;","  this.expected = options.expected;","  this.operator = options.operator;","","  if (Error.captureStackTrace) {","    Error.captureStackTrace(this, options.stackStartFunction);","  }","}","","AssertionError.prototype.__proto__ = Error.prototype;","","AssertionError.prototype.toString = function() {","  return this.message;","};"];
